art-blocks-engine-onboarding/art-blocks-engine-101/Engine-technical-details.md

```diff
@@ -0,0 +1,119 @@
+# Art Blocks Engine Flex Technical Details
+
+This page goes deeper into some technical considerations when working with Artblocks Engine Flex
+
+## Introduction To External Asset Dependencies
+
+```solidity
+struct ExternalAssetDependency {
+        string cid;
+        ExternalAssetDependencyType dependencyType;
+}
+```
+
+The Engine Flex contract introduces the concept of external asset dependencies. These essentially function as on-chain pointers to off-chain assets stored using decentralized storage technologies. An external asset dependency is comprised of its content indentifier (CID) and dependencyType, which maps to an Engine Flex supported platform. 
+
+Engine Flex currently supports adding external asset dependencies hosted using the following decentralized storage solutions:
+- IPFS
+- Arweave
+
+## Adding, Updating & Removing External Asset Dependencies
+
+When working with and manipulating a project's external asset dependencies, you'll be relying on the following functions:
+
+```solidity
+function addProjectExternalAssetDependency(uint256 _projectId, string calldata _cid, ExternalAssetDependencyType _dependencyType)
+function updateProjectExternalAssetDependency(uint256 _projectId, uint256 _index, string calldata _cid, ExternalAssetDependencyType _dependencyType)
+function removeProjectExternalAssetDependency(uint256 _projectId, uint256 _index)
+```
+
+Some important notes to keep in mind with the above functions:
+- Only allowlisted/artist addresses can call these.
+- `ExternalAssetDependencyType _dependencyType` is a solidity enum, which can be passed in to these functions as a uint8. This enum only defines two options as of now, `IPFS` and `ARWEAVE`, which can be represented as `0` and `1` respectively. 
+- In the interest of saving gas, the `removeProjectExternalAssetDependency()` function is implemented in such a way that it does not preserve the order of the project's external asset dependency mapping and, therefore, the order in which the external asset depdendencies are returned should not be relied upon in any way. Additional implementation details described here: https://github.com/ArtBlocks/artblocks-contracts/blob/main/contracts/PBAB%2BCollabs/GenArt721CoreV2_ENGINE_FLEX.sol#L596
```

Comment:
@lyaunzbe Can we chat more about this? Reading this has me realizing that I think we need to update our contract logic, as I'm not sure that the logic implemented as-is is really what we want here.

Reply:
This was discussed and decision was made to do the following:
- current logic external asset dep removal logic remains
- adding more documentation around this removal logic to ensure it's very clear
- followup: updating subgraph remove handler to check and remove empty elements in the mapping. this also sets on the path of easily forking the flex contract in the future with an updated removal function that leaves gaps in the mapping

Reply:
+++ 🙏 💜 🙏 💜 🙏 💜 🙏 💜